File: questions/q525_binary_subarrays_with_sum/optimized_sliding_window.cpp
// Single Pass Sliding Window
// (A sort of combination of both Prefix Sum and the dual sliding window approaches)

class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        int front=0, rear=0;
        int curSum = 0, pZero=0;
        int count = 0;

        while (rear < nums.size()) {
            curSum += nums[rear];

            while (curSum > goal && front < rear) {
                curSum -= nums[front++];
                pZero = 0;
            }

            while (curSum == goal && front < rear && nums[front] == 0) {
                front ++;
                pZero ++;
            }

            if (curSum == goal) {
                count += (pZero + 1);
            }
            
            rear ++;
        }

        return count;
    }
};

File: questions/q525_binary_subarrays_with_sum/optimized_sliding_window.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        front = 0
        rear = 0
        curSum = 0
        pZero = 0
        count = 0

        while (rear < len(nums)) :
            curSum += nums[rear]

            while (curSum > goal and front < rear) :
                curSum -= nums[front]
                front += 1
                pZero = 0
            
            while curSum == goal and front < rear and nums[front] == 0 :
                pZero += 1
                front += 1
            
            if curSum == goal :
                count += (1 + pZero)

            rear += 1
        
        return count
